### PR TITLE
Improve template constraint error messages

### PR DESCRIPTION
**Pull request by dayllenger**
Hello. I suppose, everyone agrees that the output on unsatisfied template constraints is very poor. This PR proposes a simple way to tell actually what's gone wrong on template matching.

Related to [issue 9626](https://issues.dlang.org/show_bug.cgi?id=9626).

### Example

This code
```D
import std.algorithm : sort;

char[] arr = ['a', 'b'];
sort(arr);
```
gives such error message:
```
./scribble.d(8): Error: template std.algorithm.sorting.sort cannot deduce function from argument types !()(char[]), candidates are:
/usr/include/dmd/phobos/std/algorithm/sorting.d(1855):        std.algorithm.sorting.sort(alias less = "a < b", SwapStrategy ss = SwapStrategy.unstable, Range)(Range r) if ((ss == SwapStrategy.unstable && (hasSwappableElements!Range || hasAssignableElements!Range) || ss != SwapStrategy.unstable && hasAssignableElements!Range) && isRandomAccessRange!Range && hasSlicing!Range && hasLength!Range)
```

"WHAT? Char array cannot be sorted?"

Now let's see how it will look with this PR:

```
./scribble.d(8): Error: template std.algorithm.sorting.sort cannot deduce function from argument types !()(char[]), candidates are:
/usr/include/dmd/phobos/std/algorithm/sorting.d(1855):        sort(alias less = "a < b", SwapStrategy ss = SwapStrategy.unstable, Range)(Range r),
  whose parameters have the following constraints:
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        ss == SwapStrategy.unstable
          > hasSwappableElements!Range
            or:
          > hasAssignableElements!Range
        or:
      > ss != SwapStrategy.unstable
      - hasAssignableElements!Range
  - isRandomAccessRange!Range
  - hasSlicing!Range
  - hasLength!Range
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  Tip: not satisfied constraints are marked with >
```

The output is structured like a tree. It shares exact information about which clauses have failed. Here we can read that our `char[]`does not have swappable and/or assignable elements, which is true, because it's UTF-8 string. Not obvious though, so thinking from the user perspective, we should move the last three checks to the beginning to improve the message.

Another (artificial) example:
```D
import std.algorithm : countUntil, initializeAll;

int[] arr;
void* i;
countUntil(arr, i);
initializeAll(i);
```
Output:
```
./scribble.d(16): Error: template std.algorithm.searching.countUntil cannot deduce function from argument types !()(int[], void*), candidates are:
/usr/include/dmd/phobos/std/algorithm/searching.d(768):        countUntil(alias pred = "a == b", R, Rs...)(R haystack, Rs needles),
  whose parameters have the following constraints:
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    isForwardRange!R
    Rs.length > 0
    isForwardRange!(Rs[0]) == isInputRange!(Rs[0])
  > is(typeof(startsWith!pred(haystack, needles[0])))
      - Rs.length == 1
        or:
      - is(typeof(countUntil!pred(haystack, needles[1..__dollar])))
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/dmd/phobos/std/algorithm/searching.d(856):        countUntil(alias pred = "a == b", R, N)(R haystack, N needle),
  whose parameters have the following constraints:
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    isInputRange!R
  > is(typeof(binaryFun!pred(haystack.front, needle)) : bool)
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/dmd/phobos/std/algorithm/searching.d(915):        countUntil(alias pred, R)(R haystack)
  Tip: not satisfied constraints are marked with >
./scribble.d(17): Error: template std.algorithm.mutation.initializeAll cannot deduce function from argument types !()(void*), candidates are:
/usr/include/dmd/phobos/std/algorithm/mutation.d(864):        initializeAll(Range)(Range range),
  whose parameters have the following constraints:
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  > isInputRange!Range
  - hasLvalueElements!Range
  - hasAssignableElements!Range
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/dmd/phobos/std/algorithm/mutation.d(913):        initializeAll(Range)(Range range),
  whose parameters have the following constraints:
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  > is(Range == char[])
    or:
  > is(Range == wchar[])
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  Tip: not satisfied constraints are marked with >
```

The output is colored. It would be prettier if I didn't care about possible breakages (e.g. it needs to move signatures that don't fit to the next line).

Why my solution is cool:
* no language changes, no DIPs
* no need to rewrite existing codebases
* may be applied to `static assert` and probably, in some way, to `static if`
* can disassemble arbitrarily complex combinations of `!`, `&&`, `||`, and `?:`

The drawback is that it is shallow, it cannot say "does not implement `popFront()`", it can only point to the constraint `isInputRange!R`, though their names are mostly quite descriptive.

Old `&&` approach to make assert messages can work here to produce some readable results, for example:
```D
void stuff(T, U)()
if (isNice!T && "must be nice" && isFine!U && "must be fine")
{}
```
will print something like
```
    isNice!T
    "must be nice"
  > isFine!U
  - "must be fine"
```

I could handle this case specially and print it better, in one line (edit: and print evaluated expression), but it's too much for this PR.

### Tools

Do we have any special output for them?

IDE can simply show that message in a tooltip. Theoretically, it can parse it and highlight failed clauses (because the message contains all parts of the constraint in order from left to right). Brittle thing, we should define a special compiler flag for error formats, and form JSON.

Then we can greatly improve the error messages for humans.

## Update

Now the full tree view with the tip message is shown only in verbose mode. In the usual, it only shows failed constraints in a compact form.

A new section was added. It clarifies what type/value the passed parameters have. It prints well usual and variadic parameters, except for lambdas (`__lambda1`), enum members (`cast(Enum)0`), maybe something else.

Examples:
```
./scribble.d(67): Error: template instance `std.array.Appender!int` does not match template declaration Appender(A)
  with A = int
  must satisfy the following constraint:
       isDynamicArray!A
```
```
./scribble.d(10): Error: template std.algorithm.sorting.sort cannot deduce function from argument types !()(FilterResult!(unaryFun, int[])), candidates are:
/usr/include/dmd/phobos/std/algorithm/sorting.d(1855):        sort(alias less = "a < b", SwapStrategy ss = SwapStrategy.unstable, Range)(Range r)
  with less = "a < b",
       ss = cast(SwapStrategy)0,
       Range = FilterResult!(unaryFun, int[])
  must satisfy the following constraint:
       isRandomAccessRange!Range
```
```
/usr/include/dmd/phobos/std/range/package.d(6264):        iota(B, E)(B begin, E end)
  with B = string,
       E = string
  must satisfy one of the following constraints:
       isIntegral!(CommonType!(B, E))
       isPointer!(CommonType!(B, E))
```
(omitted other 5 iota variants)

### CODE REVIEW

**Comment by jacob-carlborg** [comment]
This looks pretty awesome. Perhaps only print the non-failing constraints in verbose mode.

**Comment by wilzbach** [comment]
> Do we have any special output for them?

Not yet. We should really add a flag for JSON error messages. We could use the existing `-verrors`.

BTW I hope you knew about `AUTO_UPDATE=1` and didn't update all tests manually.

**Comment by dayllenger** [comment]
I updated, but before rebase. After rebase I'm getting such errors on `make -f posix.mak`:
```
dmd/backend/cg87.d:2719: error: undefined reference to 'idxregm(code*)'
dmd/backend/cgxmm.d:290: error: undefined reference to 'idxregm(code*)'
dmd/backend/cod1.d:4485: error: undefined reference to 'movOnly(elem*)'
dmd/backend/cod4.d:465: error: undefined reference to 'idxregm(code*)'
dmd/backend/cod4.d:480: error: undefined reference to 'idxregm(code*)'
collect2: error: ld returned 1 exit status
```

**Comment by wilzbach** [comment]
If you get such errors, it always helps to do a full clean.

**Comment by thewilsonator** [comment]
@dayllenger see also https://github.com/dlang/DIPs/pull/131 (will be discussed at dconf) it would be good to coordinate with these improvements (in part so that we only have to update the test suite once).

File: src/dmd/dtemplate.d [hunk 1]
```diff
@@ -537,6 +539,9 @@ extern (C++) final class TemplateDeclaration : ScopeDsymbol
     // threaded list of previous instantiation attempts on stack
     TemplatePrevious* previous;
 
+    private Expression lastConstraint; // the last failed constraint instantiation
+    private Array!Expression lastConstraintNegs; // its negative parts
```
Comment (thewilsonator) [inline src/dmd/dtemplate.d]: These should probably go in `TemplateInstance`, not declaration.

Comment (dayllenger) [reply src/dmd/dtemplate.d]: Probably, but I don't see how we can retrieve those instances for overloads in `printCandidates`.

**Comment by nordlow** [comment]
Awesome.

I would like to also have links to the failing restrictions be printed somehow so I can navigate to them inside my favorite editor or IDE.

**Comment by adamdruppe** [comment]
here's my idea for something similar btw:
https://github.com/dlang/dmd/pull/9419

One thing I like with mine is it shows what value the template parameters evaluate to, so like:

    constraint clause isRandomAccessRange!(FilterResult!(unaryFun, int[])) was false

Shows that `Range` == `FilterResult!(unaryFun, int[])`. Your source view is more readable in some cases though... but in stuff like loops, knowing which iteration failed is quite nice. Perhaps just a supplemental line that says like "You passed FilterResult!(unaryFun, int[]) as Range" (though "you passed" is semi-lie because it is the library creating that, but meh)

**Comment by dayllenger** [comment]
At first I printed those expressions after substitution, but sometimes it looks like
`is(typeof(startsWith(alias pred = (a, b) => a == b, Range, Needles...)(Range doesThisStart, Needles withOneOfThese) if (isInputRange!Range && (Needles.length > 1) && is(typeof(.startsWith!pred(doesThisStart, withOneOfThese[0])) : bool) && is(typeof(.startsWith!pred(doesThisStart, withOneOfThese[1..__dollar])) : uint))(haystack, __needles_field_0)))`...

It should be easy to add a line like `with R = int[], N = void*, pred = "a == b"`, I am already working on that.

About loops - this task is somewhat deeper than I can think over now.

**Comment by dayllenger** [comment]
See Update section in the first post. I guess, I need to put a tests pack in a file in fail_compilation?

**Comment by dayllenger** [comment]
@thewilsonator did you discuss your DIP at DConf? This PR is ready, I guess people can try it out.

**Comment by nordlow** [comment]
@dayllenger, could this be extended to include detailed transitive diagnostics of mismatched qualifiers `@safe pure nothrow @nogc` when calling template functions?

**Comment by dayllenger** [comment]
It's a good idea, and it looks complicated for me. The logic is unlikely similar: my function recursively decompose the constraint expression, but in case of function bodies it needs to intervene somehow in the mechanism of attribute inference; it requires deep knowledge of this part of the compiler I don't have currently.

File: src/dmd/dtemplate.d [hunk 2]
```diff
@@ -680,6 +686,40 @@ extern (C++) final class TemplateDeclaration : ScopeDsymbol
...
+            .toCBuffer(tp, &buf, &hgs);
+        }
+        buf.writeByte(')');
+
+        if (onemember)
+        {
+            FuncDeclaration fd = onemember.isFuncDeclaration();
+            if (fd && fd.type)
+            {
+                TypeFunction tf = cast(TypeFunction)fd.type;
```
Comment (WalterBright) [inline src/dmd/dtemplate.d]: `= fd.type.isTypeFunction();`

File: src/dmd/dtemplate.d [hunk 3]
```diff
@@ -680,6 +686,40 @@ extern (C++) final class TemplateDeclaration : ScopeDsymbol
...
+    {
+        if (literal)
+            return Dsymbol.toChars();
+
+        OutBuffer buf;
+        HdrGenState hgs;
+
+        buf.writestring(ident.toChars());
+        buf.writeByte('(');
+        for (size_t i = 0; i < parameters.dim; i++)
```
Comment (WalterBright) [inline src/dmd/dtemplate.d]: foreach (i, tp; *parameters)

File: src/dmd/dtemplate.d [hunk 4]
```diff
@@ -800,6 +848,115 @@ extern (C++) final class TemplateDeclaration : ScopeDsymbol
...
+                formatParamsWithTiargs(*lastConstraintTiargs, buf);
+                buf.writenl();
+            }
+            if (!full)
+            {
+                // choosing singular/plural
+                if (count == 1)
+                    buf.writestring("  must satisfy the following constraint:");
+                else
+                    buf.writestring("  must satisfy one of the following constraints:");
```
Comment (WalterBright) [inline src/dmd/dtemplate.d]:
```
const s = (count == 1) ? " .... " : " ... ";
buf.writestring(s);
```

File: src/dmd/errors.d [hunk 5]
```diff
@@ -400,6 +401,20 @@ extern (C++) void message(const(char)* format, ...)
     va_end(ap);
 }
 
+/**
+ * Print a tip message with the prefix and hightlighting.
```
Comment (WalterBright) [inline src/dmd/errors.d]: highlighting

File: src/dmd/staticcond.d [hunk 6]
```diff
@@ -35,82 +37,391 @@ import dmd.utils;
...
+ *      original = original expression, for error messages
  *      e =  resulting expression
  *      errors = set to `true` if errors occurred
+ *      negatives = array to store negative clauses
  * Returns:
  *      true if evaluates to true
  */
-
-bool evalStaticCondition(Scope* sc, Expression exp, Expression e, ref bool errors)
+bool evalStaticCondition(Scope* sc, Expression original, Expression e, out bool errors, Array!Expression* negatives = null)
```
Comment (WalterBright) [inline src/dmd/staticcond.d]: Expressions instead of Array!Expressions

File: src/dmd/staticcond.d [hunk 7]
```diff
@@ -35,82 +37,391 @@ import dmd.utils;
...
+                op = TOK.andAnd;
+        }
 
-    e.error("expression `%s` is not constant", e.toChars());
-    errors = true;
-    return false;
+        if (op == TOK.not)
+        {
+            NotExp no = cast(NotExp)orig;
+            NotExp ne = cast(NotExp)e;
```
Comment (WalterBright) [inline src/dmd/staticcond.d]: This is assuming that `orig` and `e` have an identical structure. Is this invariant true? How does `e` differ, then?

**Comment by WalterBright** [comment]
Given the large amount of code added over 3 files, please consider organizing this functionality into one place.

**Comment by dayllenger** [comment]
Currently the code in `staticcond.d` doesn't know anything about template constraints, so theoretically it can work with any static condition - as the module name says. The code inside `dtemplate.d` formats specifically the constraints and *also* the arguments. If I merge the parts into one, it will be unnecessary coupled.

Should I rewrite the existing commits or append a new one, for review fixes?

**Comment by JesseKPhillips** [comment]
I recommend adding fixes as new commits to review those specific changes. Use $ git commit --fixup... To be able to cleanup history quickly afterwards.

File: src/dmd/staticcond.d (hunk 7 above)
Comment (dayllenger) [reply src/dmd/staticcond.d]: Now I check `.loc` of the toplevel expressions first, so they are either equal or syntax copied. Then I check for null with assert after these casts.

**Comment by FeepingCreature** [comment]
Ping

**Comment by FeepingCreature** [comment]
Slightly more insistent ping for attention! <(｀^´)>

**Comment by dayllenger** [comment]
@WalterBright I don't see here a point of arranging the code into one place. Why it should be so? What is that place?

**Review by RazvanN7 — APPROVED** [review]

**Comment by RazvanN7** [comment]
This looks good to me. The code is logically organized, I don't see the point in trying to put it in one place. There is no reason to stall this. Adding auto-merge.

**Comment by RazvanN7** [comment]
Thank you @dayllenger for this awesome work.

**Pull request merged**